### ISSUE TITLE
Nagiosサーバをrebootするとトリガーが増加する＠15.03リリース確認

### ISSUE THREAD

**Issue by o-wakabayashi** [project-hatohol/hatohol#1163]
Nagiosを再開して、ホストグループ関連の確認をしているときに発見。

https://github.com/project-hatohol/hatohol/issues/1151　
とほぼ同じ原因と思われる。
Nagios（NDOUtils）を再開すると、トリガーIDが変わっている。
Hatoholで収集したトリガー情報のIDを見る限り、再開後は新しくトリガーIDが振られている。
（Nagios監視はアイテムもトリガーに依存するため、アイテム側も影響を受ける）

「概要：トリガー」画面はbiefが変わらないので影響がないように見える
「最新データ」「トリガー」画面は影響有り（再開する度に要素が増加する）
「イベント」はトリガーとイベントの関係が一致していれば問題ない（実質問題ない）

トリガーが増加する問題だけなら、トリガーの再読込（15.03で追加した）をすれば、
回避できるが、トリガーIDが変わるためアクションに影響がでると思う。

**Comment by o-wakabayashi** [comment]
ダッシュボードのトリガーのカウンタもおかしくなる。

**Comment by o-wakabayashi** [comment]
再開前
![screenshot- - hatohol - chromium-2](https://cloud.githubusercontent.com/assets/8002282/6818044/8722667c-d2ec-11e4-9130-5fe767bff889.png)
![screenshot- - hatohol - chromium](https://cloud.githubusercontent.com/assets/8002282/6818045/93817b92-d2ec-11e4-85c6-72f4f6251e6f.png)

mysql> select server_id,id,host_id_in_server,brief from triggers where server_id=2 and validity=1 and host_id_in_server ='machine9_1';
+-----------+------+-------------------+--------------------------------------------------------------------+
| server_id | id   | host_id_in_server | brief                                                              |
+-----------+------+-------------------+--------------------------------------------------------------------+
|         2 | 1582 | machine9_1        | PROCS OK: 29 processes with STATE = RSZDT                          |
|         2 | 1581 | machine9_1        | SWAP OK - 91% free (3714 MB out of 4093 MB)                        |
|         2 | 1580 | machine9_1        | SSH OK - OpenSSH_5.3 (protocol 2.0)                                |
|         2 | 1579 | machine9_1        | DISK OK - free space: / 4672 MB (41% inode=76%):                   |
|         2 | 1578 | machine9_1        | PING OK - Packet loss = 0%, RTA = 0.02 ms                          |
|         2 | 1577 | machine9_1        | HTTP OK: HTTP/1.1 200 OK - 430 bytes in 0.001 second response time |
|         2 | 1576 | machine9_1        | USERS OK - 0 users currently logged in                             |
|         2 | 1575 | machine9_1        | OK - load average: 0.99, 0.81, 0.81                                |
+-----------+------+-------------------+--------------------------------------------------------------------+

**Comment by o-wakabayashi** [comment]
再開後
![screenshot- - hatohol - chromium-3](https://cloud.githubusercontent.com/assets/8002282/6818049/a1ed99fe-d2ec-11e4-86da-828e27e0d412.png)
![screenshot- - hatohol - chromium-1](https://cloud.githubusercontent.com/assets/8002282/6818051/a4f008f8-d2ec-11e4-9510-da494c3a32a9.png)

mysql> select server_id,id,host_id_in_server,brief from triggers where server_id=2 and validity=1 and host_id_in_server ='machine9_1';
+-----------+------+-------------------+--------------------------------------------------------------------+
| server_id | id   | host_id_in_server | brief                                                              |
+-----------+------+-------------------+--------------------------------------------------------------------+
|         2 | 1582 | machine9_1        | PROCS OK: 29 processes with STATE = RSZDT                          |
|         2 | 1581 | machine9_1        | SWAP OK - 91% free (3714 MB out of 4093 MB)                        |
|         2 | 1580 | machine9_1        | SSH OK - OpenSSH_5.3 (protocol 2.0)                                |
|         2 | 1579 | machine9_1        | DISK OK - free space: / 4672 MB (41% inode=76%):                   |
|         2 | 1578 | machine9_1        | PING OK - Packet loss = 0%, RTA = 0.02 ms                          |
|         2 | 1577 | machine9_1        | HTTP OK: HTTP/1.1 200 OK - 430 bytes in 0.001 second response time |
|         2 | 1576 | machine9_1        | USERS OK - 0 users currently logged in                             |
|         2 | 1575 | machine9_1        | OK - load average: 0.99, 0.81, 0.81                                |
|         2 | 1621 | machine9_1        | PROCS OK: 29 processes with STATE = RSZDT                          |
|         2 | 1620 | machine9_1        | SWAP OK - 91% free (3714 MB out of 4093 MB)                        |
|         2 | 1619 | machine9_1        | SSH OK - OpenSSH_5.3 (protocol 2.0)                                |
|         2 | 1618 | machine9_1        | DISK OK - free space: / 4672 MB (41% inode=76%):                   |
|         2 | 1617 | machine9_1        | PING OK - Packet loss = 0%, RTA = 0.02 ms                          |
|         2 | 1616 | machine9_1        | HTTP OK: HTTP/1.1 200 OK - 430 bytes in 0.001 second response time |
|         2 | 1615 | machine9_1        | USERS OK - 0 users currently logged in                             |
|         2 | 1614 | machine9_1        | OK - load average: 0.99, 0.81, 0.81                                |
+-----------+------+-------------------+--------------------------------------------------------------------+

**Comment by ashie** [comment]
たしかに、取得するリソース全てに同じ問題があると思っておくべきでした。
（#1151もhostgroupだけでなくhostとhostgroup_memberに同様の問題があったのを修正している）

同じ方法で直せると思うので、直してみます。

**Comment by o-wakabayashi** [comment]
すみません、よろしくお願いします。
（前のissueの時に他のIdについても調査すべきでした）

イベントIDは加算だから問題ないのか。。。

**Comment by o-wakabayashi** [comment]
今後（15.06？）、NDOUtilsの動きも含めて対処・方針を決めないといけないですね。。。

**Comment by ashie** [comment]
service_object_idは変わらないようなので、こちらを使った方がシンプルか

**Comment by ashie** [comment]
> service_object_idは変わらないようなので、こちらを使った方がシンプルか

serviceの場合は同じ名前が存在できるようなので、これしかなさそう。

**Comment by o-wakabayashi** [comment]
nagios_servicesテーブルのservice_object_idですよね、確かにrebootしても変化しないし、
nagios_objectsテーブルのobject_idと連携するのか、各要素（オブジェクト？）を一意に
識別できそう。

2015-03-25 13:07 GMT+09:00 Takuro Ashie <notifications@github.com>:

**Comment by ashie** [comment]
> serviceの場合は同じ名前が存在できるようなので、これしかなさそう。

存在できるのではなくて、同じ名前の設定があると内容が上書きされるようだ。

**Comment by ashie** [comment]
ndoutilsのコードを確認したが、やはりnagios再起動時に今回問題となっているテーブルを
クリアする処理がある。Ubuntu 14.04のndoutilsを使用しているので、バージョンは1.4b9。

src/dbhandlers.c:
```C
        /* if process is starting up, clearstatus data, event queue, etc. */
        if(type==NEBTYPE_PROCESS_PRELAUNCH && tstamp.tv_sec>=idi->dbinfo.latest_realtime_data_time){

                ...
                ndo2db_db_clear_table(idi,ndo2db_db_tablenames[NDO2DB_DBTABLE_HOSTGROUPS]);
                ndo2db_db_clear_table(idi,ndo2db_db_tablenames[NDO2DB_DBTABLE_HOSTGROUPMEMBERS]);
                ...
                ndo2db_db_clear_table(idi,ndo2db_db_tablenames[NDO2DB_DBTABLE_SERVICES]);
                ...
```

src/db.c:
```C
int ndo2db_db_clear_table(ndo2db_idi *idi, char *table_name){
        ...
        if(asprintf(&buf,"DELETE FROM %s WHERE instance_id='%lu'"
```

**Comment by ashie** [comment]
nagios_objectsはクリアされない。

**Comment by ashie** [comment]
nagiosの終了時ではなく、開始時に上記処理が走る。

**Comment by ashie** [comment]
nagios_objectsをクリアしている箇所はないし、テンプレートを使っても同じobject_idが重複して使われるということはない(serviceの場合はname1がhost_name、name2がserv ice_descriptionになる)。
オブジェクトが消された場合は、nagios_objectsからdeleteされるのではなくて、in_activeが1になる。

テーブルjoinした上でname1とname2を組み合わせて重複の無いようにエスケープするのは面倒なので、同じ単位で一意性のあるobject_idを使うのが簡単だし妥当そう。

この方法で特に問題なく動いている（アイテムとトリガーは増えなくなった）。

**Comment by ashie** [comment]
イベントは単調増加(AUTO_INCREMENT)だし、hatohol側でも前回のIDを覚えているので、重複することはない。

ちなみにIDが最大値に達した場合、MySQLでは行を追加できなくなる。
ndoutilsでIDを巻き戻している箇所もない。

**Comment by ashie** [comment]
以上で全ての種類のリソースを網羅できているはず。

**Comment by o-wakabayashi** [comment]
ホスト関連のID（メンバー/グループなど）は#1151で修正していただいた。
イベントIDは発生したらインクリメントなので、重複しない。
トリガーIDはコレで修正していただいた。 ★
アイテム（最新データ）はトリガーと同じIDを使用するため、コノ対処と同様。 ★

今回は★について修正していただいた。

ID意識するのは、ホスト関連以外だと、
actions ：トリガーID
events ：トリガーID/イベントID
triggers ：トリガーID
items：アイテムID（NagiosではトリガーIDと同意）
だと思いますが、認識あいますか？。


2015-03-25 16:26 GMT+09:00 Takuro Ashie <notifications@github.com>:

**Comment by ashie** [comment]
はい、同じ認識です。

**Comment by o-wakabayashi** [comment]
はい、確認ありがとうございます。

2015-03-25 16:37 GMT+09:00 Takuro Ashie <notifications@github.com>:

**Comment by o-wakabayashi** [comment]
Thank you for closed of issue.
